Mathlib/Tactic/Congrm.lean

```diff
@@ -0,0 +1,137 @@
+    MetaM (List MVarId) := do
+  goal.apply (← mkConstWithFreshMVarLevels lem)
+
+end util
+
+private partial def telescopingFn (lem : Name) (goal : MVarId) (x : Expr) : MetaM MVarId := do
+  let userName ← x.fvarId!.getUserName
+  let newGoals ← goal.applyWithFreshMVarLevels lem
+  if newGoals.length == 1 then
+    let newGoal := newGoals[0]!
+    return (← newGoal.intro userName).2
+  throwTacticEx `congrm goal m!"failed to apply {lem}"
+
+open private applyCongrThm? from Lean.Meta.Tactic.Congr in
+
+/-- Main loop for `congrm` -/
+partial def congrmLoop (pat : Expr) (goal : MVarId) : MetaM (List MVarId) := do
+  -- Helper function (stolen from somewhere) that creates the correct FVars in `λ` and `∀`
+  -- and does the recursion
+  let binders (mvarId : MVarId) (n : Name) (xs : Array Expr) (k : Expr) : MetaM (List MVarId) := do
+    congrmLoop k (← xs.foldlM (telescopingFn n) mvarId)
+  if pat.isMVar then
+    return [goal]
+  else if pat.isForall then
+    trace[congrm] s!"Forall pattern: {← ppExpr pat}"
+    forallTelescope pat (binders goal ``pi_congr)
+  else if pat.isLambda then
+    trace[congrm] s!"Lambda pattern: {← ppExpr pat}"
+    lambdaTelescope pat (binders goal ``funext)
+  else if pat.isApp then
+    let patternArgs := (← pat.getExplicitArgs).toList
+    trace[congrm] s!"Apply pattern, fun: {← ppExpr pat.getAppFn}, args: {← patternArgs.mapM ppExpr}"
+    let some congrThm ← mkCongrSimp? pat.getAppFn' (subsingletonInstImplicitRhs := false) |
+      return []
+    if pat.getAppFn.isMVar then
+      -- If the function is a metavariable, we just return the goal
+      return [goal]
+
+    let goalArgs ← applyCongrThm? goal congrThm
+    -- If the pattern has a different number of arguments, then we just fail:
+    if goalArgs.length != patternArgs.length then
+      trace[congrm] s!"Number of patterns and arguments are different:
+        args: {goalArgs.map MVarId.name}
+        pat: {← patternArgs.mapM ppExpr}"
+      throwTacticEx `congrm goal m!"Number of function arguments does not match"
+
+    -- Apply `congrm_loop` to all arguments with the corresponding pattern and concat the resulting
+    -- list.
+    return (← (patternArgs.zip goalArgs).mapM (λ (e,m) => congrmLoop e m)).join
+  else if pat.isFVar then
+    let fvarId := pat.fvarId!
+    trace[congrm] s!"FVar: {← fvarId.getUserName}"
+    let res ← observing? do
+      goal.refl
+    if res == none then
+      throwTacticEx `congrm goal m!"Variable pattern is not proved by `refl`"
+    else return []
+  else
+    trace[congrm] s!"We have an unhandled expression: {← ppExpr pat}"
+    throwTacticEx `congrm goal m!"Invalid pattern"
+
+/-- the core function for `congrm` -/
+partial def congrmCore (pat : Expr) (goal : MVarId) : MetaM (List MVarId) := do
+  -- First change `iff` to `=` and then run the loop:
+  let mvars ← congrmLoop pat (← goal.iffOfEq)
+  -- Try `refl` on all remaining goals
+  mvars.filterMapM tryRefl
```

Comment:
You're allowed to return assigned metavariables with `liftMetaTactic`, so you can do something like this, if we want to avoid needing `tryRefl`:
```lean
for mvar in mvars do
  try discard <| mvar.refl catch _ => pure ()
return mvars
```
Whenever you get the main goal, it prunes assigned metavariables for you anyway.